# ML-03 — Frame Your Lane as an ML Task

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/BrightTimothy/FlyRank-AI-Machine-Learning-Internship/blob/main/work/notebooks/w02_ml_task_framing.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane as an ML task (type)

This is fundamentally a scoring/ranking task: rather than sorting pages into clean categories or grouping similar ones, the output is a continuous score per page, used to rank all visible pages by how much they under-capture clicks or engagement relative to peers. A binary classification layer can sit underneath it (flagged opportunity vs. not) to help calibrate the score, but the deliverable itself is a ranked queue, not a single yes/no per page.

In [3]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 2. Target or proxy

There's no ground-truth label for "this page has a fixable CTR problem" — that would require running an actual fix and observing the outcome, which this data doesn't give me. So I use a proxy: a page's CTR gap relative to its position tier's average CTR (ctr_gap = observed_ctr − expected_ctr_for_tier). A large negative gap is the proxy for "opportunity." It's a proxy, not ground truth, because a low gap could have other causes I can't rule out from this data alone (seasonality, wrong intent match, etc.).

In [4]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 3. Success metric

Precision@50 — because the reviewer can only realistically check a fixed number of pages per week (established in ML-02, section 2), what matters isn't overall accuracy across all pages, it's whether the top of the ranked list is actually worth their limited time. A model with mediocre overall accuracy but a sharp top-50 is more useful here than the reverse.

In [5]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 4. The unit of analysis, as a real dataframe

*Load your lane's slice and show it: one row = one what?*

In [6]:
import pandas as pd

df = pd.read_csv("https://raw.githubusercontent.com/BrightTimothy/FlyRank-AI-Machine-Learning-Internship/main/data/raw/content_refresh_anonymized.csv")

lane4_cols = ["content_id", "impressions_90d", "clicks_90d", "ctr", "avg_position",
              "position_tier", "sessions_90d", "engagement_rate", "scroll_rate", "content_age_days"]
lane4_df = df[lane4_cols]
lane4_df.head()

expected_ctr = df.groupby("position_tier")["ctr"].transform("mean")
lane4_df["ctr_gap"] = df["ctr"] - expected_ctr

lane4_df[["content_id", "position_tier", "ctr", "ctr_gap"]].sort_values("ctr_gap").head()


/tmp/ipykernel_2253/1026442516.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  lane4_df["ctr_gap"] = df["ctr"] - expected_ctr


,content_id,position_tier,ctr,ctr_gap
18444,content_930f8842e0d3,top_3,0.0,-1.483611
4345,content_119438e47da5,top_3,0.0,-1.483611
4344,content_031a458bf3c8,top_3,0.0,-1.483611
4341,content_925d5c673b05,top_3,0.0,-1.483611
18597,content_9e0e72b55bf4,top_3,0.0,-1.483611


## 5. Why ML beats a fixed rule here

A fixed rule (like ctr < 0.5) ignores that expected CTR varies nearly 10x across position tiers (0.15 in the deepest tier vs. 1.48 in top-3, from ML-02's real numbers) — so one flat threshold will over-flag deep pages and under-flag top-tier pages that are still genuinely underperforming for their tier. A learned scoring approach can weigh position, engagement, and volume signals together and adapt the threshold implicitly per tier, instead of me guessing one number that has to work everywhere at once.

In [7]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.